ledger/ledgercore/persistedacctdata.go

```diff
@@ -116,3 +117,458 @@ var EncodedMaxAppLocalStates = basics.EncodedMaxAppLocalStates
+func (g *AssetsHoldingGroup) Update(ai int, holdings basics.AssetHolding) {
+	g.groupData.update(ai, holdings)
+}
+
+// Loaded return a boolean flag indicated if the group loaded or not
+func (g AssetsHoldingGroup) Loaded() bool {
+	return g.loaded
+}
+
+// Load sets a group data value in the group
+func (g *AssetsHoldingGroup) Load(gd AssetsHoldingGroupData) {
+	g.groupData = gd
+	g.loaded = true
+}
+
+func (g *AssetsHoldingGroup) delete(ai int) {
+	// although a group with only one element is handled by a caller
+	// add a safety check here
+	if g.Count == 1 {
+		*g = AssetsHoldingGroup{}
+		return
+	}
+
+	if ai == 0 {
+		// when deleting the first element, update MinAssetIndex
+		g.MinAssetIndex += g.groupData.AssetOffsets[1]
+	} else if uint32(ai) == g.Count-1 {
+		// when deleting the last element, update DeltaMaxAssetIndex
+		g.DeltaMaxAssetIndex -= uint64(g.groupData.AssetOffsets[len(g.groupData.AssetOffsets)-1])
+	}
+	g.groupData.delete(ai)
+	g.Count--
+}
+
+func (g *AssetsHoldingGroup) insert(aidx basics.AssetIndex, holding basics.AssetHolding) {
+	if aidx < g.MinAssetIndex {
+		// prepend
+		g.groupData.Amounts = append([]uint64{holding.Amount}, g.groupData.Amounts...)
+		g.groupData.Frozens = append([]bool{holding.Frozen}, g.groupData.Frozens...)
+		g.groupData.AssetOffsets[0] = g.MinAssetIndex - aidx
+		g.groupData.AssetOffsets = append([]basics.AssetIndex{0}, g.groupData.AssetOffsets...)
+		g.MinAssetIndex = aidx
+	} else if aidx >= g.MinAssetIndex+basics.AssetIndex(g.DeltaMaxAssetIndex) {
+		// append
+		g.groupData.Amounts = append(g.groupData.Amounts, holding.Amount)
+		g.groupData.Frozens = append(g.groupData.Frozens, holding.Frozen)
+		lastAssetIndex := g.MinAssetIndex + basics.AssetIndex(g.DeltaMaxAssetIndex)
+		delta := aidx - lastAssetIndex
+		g.groupData.AssetOffsets = append(g.groupData.AssetOffsets, delta)
+		g.DeltaMaxAssetIndex = uint64(aidx - g.MinAssetIndex)
+	} else {
+		// find position and insert
+		cur := g.MinAssetIndex
+		for ai, d := range g.groupData.AssetOffsets {
+			cur = d + cur
+			if aidx < cur {
+				g.groupData.AssetOffsets = append(g.groupData.AssetOffsets, 0)
+				copy(g.groupData.AssetOffsets[ai:], g.groupData.AssetOffsets[ai-1:])
+				prev := cur - d
+				g.groupData.AssetOffsets[ai] = aidx - prev
+				g.groupData.AssetOffsets[ai+1] = cur - aidx
+
+				g.groupData.Amounts = append(g.groupData.Amounts, 0)
+				copy(g.groupData.Amounts[ai:], g.groupData.Amounts[ai-1:])
+				g.groupData.Amounts[ai] = holding.Amount
+
+				g.groupData.Frozens = append(g.groupData.Frozens, false)
+				copy(g.groupData.Frozens[ai:], g.groupData.Frozens[ai-1:])
+				g.groupData.Frozens[ai] = holding.Frozen
+
+				break
+			}
+		}
+	}
+	g.Count++
+}
+
+// Delete removes asset by index ai from group gi.
+// It returns true if the group gone and needs to be removed from DB
+func (e *ExtendedAssetHolding) Delete(gi int, ai int) bool {
+	if e.Groups[gi].Count == 1 {
+		copy(e.Groups[gi:], e.Groups[gi+1:])
```

Comment:
you might want to write an explicit clean here to the last entry so that the GC could release it's arrays.

Reply:
exactly.